### DISCUSSION TITLE
[Feature Request] Support for CSS container queries

### DISCUSSION THREAD

**Discussion by martinszeltins** [tailwindlabs/tailwindcss#5482] [Ideas]
Since CSS container queries are already available in Chrome behind a flag, is this something that will be coming to Tailwind CSS soon as well?

After all, as it says on tailwindcss.com

> MODERN FEATURES
> Cutting-edge is our comfort zone.

**Comment by martinszeltins** [comment]
Maybe for version 4? 🤞🏻

**Comment by VicGUTT** [comment]
Yeah, was curious about it as well.
Wanted to start a [discussion on the subject a while back](https://github.com/tailwindlabs/tailwindcss/discussions/4529) and WIPed up a small proof of concept that I never published _(along with many more Tailwind related plugins ^^)_.

We shall see what the future holds :v:

**Comment by adamwathan** [comment]
Yeah definitely something I'd like to play with in the future — I tweeted about some API ideas back in June:

https://twitter.com/adamwathan/status/1404795418651856903

I think I'd like to figure out "arbitrary variants" as a universal concept first, then something like this would be pretty straightforward 👍🏻

**Comment by martinszeltins** [comment]
We now have a polyfill available so we can already start using container queries. It would be so nice to have Tailwind support for this. For now we need to write custom css though.

https://css-tricks.com/a-new-container-query-polyfill-that-just-works/
https://github.com/GoogleChromeLabs/container-query-polyfill

**Comment by MichaelAllenWarner** [comment]
Here's an idea for an implementation (perhaps not the _only_ implementation, and perhaps only enabled optionally): if possible, adjust the selectors for the normal responsive variants (`lg:` and the like) so that they become container-query variants when they appear within a container-query ancestor. This way, it becomes extremely easy to build a component that can appear either as a full-width piece or against a sidebar.

(I'm actually already doing something like this with a custom PostCSS plugin and a client-side script, though it's a `ResizeObserver` solution, not a true container-query solution.)

**Comment by martinszeltins** [comment]
I have found this plugin that seems to work ok - https://github.com/dgknca/tailwindcss-container-query

I like @MichaelAllenWarner suggestion but I think we will need a lot more sizes than sm, md, lg, xl and 2xl

**Reply by MichaelAllenWarner** [reply]
Well, you can always add more responsive variants to `screens` in your Tailwind config. My current implementation already allows for that (both my custom PostCSS plugin and my client-side script use the Tailwind config to generate their list of breakpoints).

In case anyone's curious...

The PostCSS plugin works by taking each responsive rule like:

```css
@media (min-width: 640px) {
  .sm\:mt-10 {
    margin-top: 2.5rem;
  }
}
```

and making a new rule out of it like:

```css
._sm ._sm\:mt-10 {
  margin-top: 2.5rem;
}
```

The client-side script is based on [this `ResizeObserver` implementation](https://philipwalton.com/articles/responsive-components-a-solution-to-the-container-queries-problem/). It sets up the ResizeObservers for the container-query ancestors (which I flag with a particular data-attribute), and it loops through the descendants of those elements to prefix every responsive variant it finds with an underscore (so that `class="sm:mt-10"` becomes `class="_sm:mt-10"` in the actual DOM). It also uses a `MutationObserver` to handle dynamically added elements after page-load.

So I only have to use my normal Tailwind responsive variants in development. Anything that appears within an element that's flagged as a container-query ancestor automatically gets the container-query treatment.

Some limitations:

- Only works with JavaScript enabled in the browser (but so do all container-query polyfills).
- Only works with the responsive classes actually in the markup (no `@apply`, etc., but I prefer to put everything in the markup anyway, so this is rarely a problem for me).
  - And doesn't work with 3rd-party styles, of course.
- No nesting container-query ancestors unless you make the selector a bit more complex (which I do, so I do get one extra "layer" of nesting).
  - Relatedly, the container-query ancestors aren't named, so the nesting mechanism simply prevents the outer ancestor's width from affecting descendants that are within an inner ancestor.
- Using the same breakpoints for media-query and container-query isn't always ideal, especially if you're also using [container-components](https://tailwindcss.com/docs/container) that apply max-widths.

But overall it works really well for my purposes, the primary one being making pieces that are styled appropriately either as full-width slabs or against a sidebar (without having to go out of my way to handle these as separate cases).

**Comment by martinszeltins** [comment]
From my experience, the breakpoints for container queries need a much more granular approach than for viewport queries. For example if you are targeting a container inside a grid like a calendar day. It almost seems that there needs to be a dynamic (per case) breakpoints instead of the more general (sm, md, lg) - since they don't seem to be very applicable for the vast amount of container sizes.

Thoughts?
